File: src/stacksniff/analyzers/fingerprint_matcher.py
"""Match evidence against technology fingerprints."""

from __future__ import annotations

import logging
import re
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from stacksniff.fingerprints import FingerprintStore
from stacksniff.models import Evidence, TechMatch

logger = logging.getLogger(__name__)


def _normalize_js_key(k: str) -> str:
    """Normalize a JS global variable name for matching.

    Removes 'window.' prefix, handles optional chaining, and lowercases.
    """
    if k.startswith("window."):
        k = k[7:]
    k = k.replace("?.", ".")
    if k.endswith("()"):
        k = k[:-2]
    return k.lower().strip()


class FingerprintMatcher:
    """Matches collected evidence against technology fingerprints."""

    def __init__(self, store: FingerprintStore) -> None:
        self.store = store

    def match(self, evidence_data: dict[str, Any]) -> list[TechMatch]:
        """Match collected evidence against all fingerprints in the store.

        Parameters
        ----------
        evidence_data:
            The raw evidence dictionary containing:
            - headers (dict[str, str])
            - cookies (dict[str, str])
            - meta_tags (dict[str, str])
            - script_srcs (list[str])
            - link_hrefs (list[str])
            - html (str or raw_html)
            - js_globals (dict[str, str])

        Returns
        -------
        list[TechMatch]
            List of matched technologies sorted by confidence descending.
        """
        # Extract fields with safe fallbacks
        headers = {k.lower(): v for k, v in evidence_data.get("headers", {}).items()}
        cookies = {k.lower(): v for k, v in evidence_data.get("cookies", {}).items()}
        meta_tags = {k.lower(): v for k, v in evidence_data.get("meta_tags", {}).items()}
        script_srcs = list(evidence_data.get("script_srcs", []))
        link_hrefs = list(evidence_data.get("link_hrefs", []))
        raw_html = str(evidence_data.get("html", "") or evidence_data.get("raw_html", ""))
        js_globals = {
            _normalize_js_key(k): v for k, v in evidence_data.get("js_globals", {}).items()
        }

        # Combine scripts and link hrefs for broader script matching
        all_scripts = script_srcs + link_hrefs

        matches: list[TechMatch] = []

        for fp in self.store.get_all():
            matched_sources: set[str] = set()
            evidences: list[Evidence] = []
            versions: list[str] = []

            # 1. Match Headers
            for header_key, pattern in fp.headers.items():
                header_val = headers.get(header_key.lower())
                if header_val is not None:
                    try:
                        rx = re.compile(pattern, re.IGNORECASE)
                        match = rx.search(header_val)
                        if match:
                            matched_sources.add("header")
                            evidences.append(
                                Evidence(
                                    source="header",
                                    key=header_key,
                                    matched=header_val,
                                    pattern=pattern,
                                )
                            )
                            if rx.groups > 0 and len(match.groups()) >= 1:
                                v = match.group(1)
                                if v:
                                    versions.append(v)
                    except re.error as exc:
                        logger.warning(
                            "Invalid regex in header rule for %s -> %s: %s",
                            fp.name,
                            header_key,
                            exc,
                        )

            # 2. Match Cookies
            for cookie_key, pattern in fp.cookies.items():
                cookie_val = cookies.get(cookie_key.lower())
                if cookie_val is not None:
                    try:
                        rx = re.compile(pattern, re.IGNORECASE)
                        match = rx.search(cookie_val)
                        if match:
                            matched_sources.add("cookie")
                            evidences.append(
                                Evidence(
                                    source="cookie",
                                    key=cookie_key,
                                    matched=cookie_val,
                                    pattern=pattern,
                                )
                            )
                            if rx.groups > 0 and len(match.groups()) >= 1:
                                v = match.group(1)
                                if v:
                                    versions.append(v)
                    except re.error as exc:
                        logger.warning(
                            "Invalid regex in cookie rule for %s -> %s: %s",
                            fp.name,
                            cookie_key,
                            exc,
                        )

            # 3. Match Meta Tags
            for meta_key, pattern in fp.meta.items():
                meta_val = meta_tags.get(meta_key.lower())
                if meta_val is not None:
                    try:
                        rx = re.compile(pattern, re.IGNORECASE)
                        match = rx.search(meta_val)
                        if match:
                            matched_sources.add("meta")
                            evidences.append(
                                Evidence(
                                    source="meta",
                                    key=meta_key,
                                    matched=meta_val,
                                    pattern=pattern,
                                )
                            )
                            if rx.groups > 0 and len(match.groups()) >= 1:
                                v = match.group(1)
                                if v:
                                    versions.append(v)
                    except re.error as exc:
                        logger.warning(
                            "Invalid regex in meta rule for %s -> %s: %s",
                            fp.name,
                            meta_key,
                            exc,
                        )

            # 4. Match Script Src / Link Hrefs
            for pattern in fp.scripts:
                matched_script = None
                matched_val = ""
                # Try substring check first for speed
                for script in all_scripts:
                    if pattern in script:
                        matched_script = script
                        matched_val = script
                        break

                # Fallback to full regex matching
                if not matched_script:
                    try:
                        rx = re.compile(pattern, re.IGNORECASE)
                        for script in all_scripts:
                            match = rx.search(script)
                            if match:
                                matched_script = script
                                matched_val = script
                                if rx.groups > 0 and len(match.groups()) >= 1:
                                    v = match.group(1)
                                    if v:
                                        versions.append(v)
                                break
                    except re.error as exc:
                        logger.warning(
                            "Invalid regex in script rule for %s -> %s: %s",
                            fp.name,
                            pattern,
                            exc,
                        )

                if matched_script:
                    matched_sources.add("script")
                    evidences.append(
                        Evidence(
                            source="script",
                            key="script_src",
                            matched=matched_val,
                            pattern=pattern,
                        )
                    )

            # 5. Match HTML
            for pattern in fp.html:
                try:
                    rx = re.compile(pattern, re.IGNORECASE)
                    match = rx.search(raw_html)
                    if match:
                        matched_sources.add("html")
                        evidences.append(
                            Evidence(
                                source="html",
                                key="raw_html",
                                matched=match.group(0)[:100] + "...",
                                pattern=pattern,
                            )
                        )
                        if rx.groups > 0 and len(match.groups()) >= 1:
                            v = match.group(1)
                            if v:
                                versions.append(v)
                except re.error as exc:
                    logger.warning(
                        "Invalid regex in html rule for %s -> %s: %s",
                        fp.name,
                        pattern,
                        exc,
                    )

            # 6. Match JS Globals
            for global_key, pattern in fp.js_globals.items():
                normalized_rule_key = _normalize_js_key(global_key)
                # Check for existence of the normalized key
                if normalized_rule_key in js_globals:
                    global_val = js_globals[normalized_rule_key]
                    matched_js = False
                    if pattern == ".":
                        matched_js = True
                    else:
                        try:
                            rx = re.compile(pattern, re.IGNORECASE)
                            match = rx.search(global_val)
                            if match:
                                matched_js = True
                                if rx.groups > 0 and len(match.groups()) >= 1:
                                    v = match.group(1)
                                    if v:
                                        versions.append(v)
                        except re.error as exc:
                            logger.warning(
                                "Invalid regex in js_global rule for %s -> %s: %s",
                                fp.name,
                                global_key,
                                exc,
                            )

                    if matched_js:
                        matched_sources.add("js_global")
                        truncated_val = (
                            global_val[:100] + "..." if len(global_val) > 100 else global_val
                        )
                        evidences.append(
                            Evidence(
                                source="js_global",
                                key=global_key,
                                matched=truncated_val,
                                pattern=pattern,
                            )
                        )

            # 7. Match DOM selectors
            dom_data = evidence_data.get("dom", {})
            if fp.dom:
                if isinstance(fp.dom, list):
                    for sel in fp.dom:
                        if sel in dom_data:
                            matched_sources.add("dom")
                            matched_el = dom_data[sel][0]
                            matched_text = f"Found element: {sel}"
                            if matched_el.get("text"):
                                matched_text += f" (text: {matched_el['text']})"
                            evidences.append(
                                Evidence(
                                    source="dom",
                                    key="selector",
                                    matched=matched_text[:100],
                                    pattern=sel,
                                )
                            )
                elif isinstance(fp.dom, dict):
                    for sel, sub_rule in fp.dom.items():
                        if sel in dom_data:
                            elements = dom_data[sel]
                            if not sub_rule or "exists" in sub_rule:
                                matched_sources.add("dom")
                                matched_text = f"Element exists: {sel}"
                                evidences.append(
                                    Evidence(
                                        source="dom",
                                        key=sel,
                                        matched=matched_text,
                                        pattern="",
                                    )
                                )
                                continue

                            matched_any_el = False
                            for el in elements:
                                matched_el_criteria = True

                                # Check text regex
                                if "text" in sub_rule:
                                    pat = sub_rule["text"]
                                    try:
                                        rx = re.compile(pat, re.IGNORECASE)
                                        match = rx.search(el.get("text", ""))
                                        if not match:
                                            matched_el_criteria = False
                                        else:
                                            if rx.groups > 0 and len(match.groups()) >= 1:
                                                v = match.group(1)
                                                if v:
                                                    versions.append(v)
                                    except re.error:
                                        matched_el_criteria = False

                                # Check attributes
                                if matched_el_criteria and "attributes" in sub_rule:
                                    el_attrs = el.get("attributes", {})
                                    for attr_name, pat in sub_rule["attributes"].items():
                                        attr_val = el_attrs.get(attr_name)
                                        if attr_val is None:
                                            matched_el_criteria = False
                                            break
                                        try:
                                            rx = re.compile(pat, re.IGNORECASE)
                                            match = rx.search(attr_val)
                                            if not match:
                                                matched_el_criteria = False
                                                break
                                            else:
                                                if rx.groups > 0 and len(match.groups()) >= 1:
                                                    v = match.group(1)
                                                    if v:
                                                        versions.append(v)
                                        except re.error:
                                            matched_el_criteria = False
                                            break

                                # Check properties
                                if matched_el_criteria and "properties" in sub_rule:
                                    el_props = el.get("properties", {})
                                    for prop_name, pat in sub_rule["properties"].items():
                                        prop_val = el_props.get(prop_name)
                                        if prop_val is None:
                                            matched_el_criteria = False
                                            break
                                        try:
                                            rx = re.compile(pat, re.IGNORECASE)
                                            match = rx.search(prop_val)
                                            if not match:
                                                matched_el_criteria = False
                                                break
                                            else:
                                                if rx.groups > 0 and len(match.groups()) >= 1:
                                                    v = match.group(1)
                                                    if v:
                                                        versions.append(v)
                                        except re.error:
                                            matched_el_criteria = False
                                            break

                                if matched_el_criteria:
                                    matched_any_el = True
                                    matched_details = []
                                    if "text" in sub_rule:
                                        matched_details.append(f"text: {el.get('text')}")
                                    if "attributes" in sub_rule:
                                        matched_details.append(
                                            f"attributes: { {k: el.get('attributes', {}).get(k) for k in sub_rule['attributes']} }"
                                        )
                                    matched_str = (
                                        f"Found: {sel} ({', '.join(matched_details)})"
                                        if matched_details
                                        else f"Found: {sel}"
                                    )
                                    evidences.append(
                                        Evidence(
                                            source="dom",
                                            key=sel,
                                            matched=matched_str[:120],
                                            pattern=str(sub_rule),
                                        )
                                    )
                                    break

                            if matched_any_el:
                                matched_sources.add("dom")

            # Calculate final confidence and build match
            if evidences:
                num_sources = len(matched_sources)
                confidence = fp.confidence

                # +0.1 per additional corroborating source type
                if num_sources > 1:
                    confidence += 0.1 * (num_sources - 1)

                # +0.1 if version capture group matched
                version = versions[0] if versions else None
                if version:
                    confidence += 0.1

                confidence = min(confidence, 1.0)

                matches.append(
                    TechMatch(
                        name=fp.name,
                        category=fp.category,
                        version=version,
                        confidence=round(confidence, 2),
                        evidence=evidences,
                    )
                )

        # Resolve implies chains
        matches = self.store.resolve_implies(matches)

        # Sort matches by confidence descending
        matches.sort(key=lambda m: m.confidence, reverse=True)

        return matches

File: stacksniff-ui/src/matchers/FingerprintMatcher.ts
// src/matchers/FingerprintMatcher.ts — ported from src/stacksniff/analyzers/fingerprint_matcher.py

import { CollectedEvidence, TechMatch, Evidence } from "../types/evidence";

export interface Fingerprint {
  name: string;
  category: string;
  website?: string | null;
  headers: Record<string, string>;
  cookies: Record<string, string>;
  meta: Record<string, string>;
  scripts: string[];
  html: string[];
  js_globals: Record<string, string>;
  dom: Record<string, any> | string[];
  implies: string[];
  confidence: number;
}

export class FingerprintStore {
  version: string;
  categories: Record<string, Record<string, string>>;
  technologies: Map<string, Fingerprint>;

  constructor(rawData: any) {
    this.version = String(rawData.version || "1.0.0");
    this.categories = rawData.categories || {};
    this.technologies = new Map<string, Fingerprint>();

    const rawTechs = rawData.technologies || {};
    for (const [key, data] of Object.entries(rawTechs)) {
      if (!data || typeof data !== "object") continue;
      const t = data as any;
      const name = String(t.name || key);
      const catSlug = String(t.category || "other");
      const catInfo = this.categories[catSlug];
      const categoryName = catInfo ? String(catInfo.name || catSlug) : catSlug;

      const fingerprint: Fingerprint = {
        name,
        category: categoryName,
        website: t.website ? String(t.website) : null,
        headers: t.headers || {},
        cookies: t.cookies || {},
        meta: t.meta || {},
        scripts: Array.isArray(t.scripts) ? t.scripts.map(String) : [],
        html: Array.isArray(t.html) ? t.html.map(String) : [],
        js_globals: t.js_globals || {},
        dom: Array.isArray(t.dom) ? t.dom.map(String) : (t.dom || {}),
        implies: Array.isArray(t.implies) ? t.implies.map(String) : [],
        confidence: typeof t.confidence === "number" ? t.confidence : 0.5,
      };

      this.technologies.set(key.toLowerCase(), fingerprint);
    }
  }

  get_all(): Fingerprint[] {
    return Array.from(this.technologies.values());
  }

  get_all_dom_selectors(): Set<string> {
    const selectors = new Set<string>();
    for (const f of this.technologies.values()) {
      if (Array.isArray(f.dom)) {
        for (const sel of f.dom) selectors.add(sel);
      } else if (f.dom && typeof f.dom === "object") {
        for (const sel of Object.keys(f.dom)) selectors.add(sel);
      }
    }
    return selectors;
  }

  resolve_implies(matches: TechMatch[]): TechMatch[] {
    const matchMap = new Map<string, TechMatch>();
    for (const m of matches) {
      matchMap.set(m.name.toLowerCase(), m);
    }

    const queue = Array.from(matchMap.keys());
    const visited = new Set<string>(queue);

    while (queue.length > 0) {
      const currentTech = queue.shift()!;
      const rule = this.technologies.get(currentTech);
      if (!rule || !rule.implies) {
        continue;
      }

      for (const implied of rule.implies) {
        const impliedKey = implied.toLowerCase();
        const impliedRule = this.technologies.get(impliedKey);
        const impliedName = impliedRule ? impliedRule.name : implied;
        const impliedCategory = impliedRule ? impliedRule.category : "";

        const existing = matchMap.get(impliedKey);
        if (existing) {
          if (existing.confidence < 0.6) {
            matchMap.set(impliedKey, {
              name: existing.name,
              category: existing.category,
              version: existing.version,
              confidence: 0.6,
              evidence: [
                ...existing.evidence,
                {
                  source: "implies",
                  key: "implies",
                  matched: `Implied by ${rule.name}`,
                  pattern: "",
                },
              ],
            });
          }
        } else {
          matchMap.set(impliedKey, {
            name: impliedName,
            category: impliedCategory,
            version: null,
            confidence: 0.6,
            evidence: [
              {
                source: "implies",
                key: "implies",
                matched: `Implied by ${rule.name}`,
                pattern: "",
              },
            ],
          });
        }

        if (!visited.has(impliedKey)) {
          visited.add(impliedKey);
          queue.push(impliedKey);
        }
      }
    }

    return Array.from(matchMap.values());
  }
}

export function _normalize_js_key(k: string): string {
  let key = k.trim();
  if (key.startsWith("window.")) {
    key = key.slice(7);
  }
  key = key.replace(/\?\./g, ".");
  if (key.endsWith("()")) {
    key = key.slice(0, -2);
  }
  return key.toLowerCase().trim();
}

const SELECTOR_PATTERN = /^(?<tag>[a-zA-Z0-9_-]*)\[(?<attr>[a-zA-Z0-9_-]+)(?<op>[~|^$*]?=)['"]?(?<val>[^'"\]]+)['"]?\]$/;

export function _match_network_request_to_selector(url: string, sel: string): any | null {
  const match = SELECTOR_PATTERN.exec(sel);
  if (!match || !match.groups) {
    return null;
  }

  const { attr, op, val } = match.groups;

  if (attr !== "href" && attr !== "src" && attr !== "data-href") {
    return null;
  }

  let matched = false;
  if (op === "=") {
    matched = url === val || (
      !val.startsWith("http:") && !val.startsWith("https:") && !val.startsWith("//") && url.endsWith(val)
    );
  } else if (op === "*=") {
    matched = url.includes(val);
  } else if (op === "^=") {
    matched = url.startsWith(val);
  } else if (op === "$=") {
    matched = url.endsWith(val);
  } else if (op === "~=") {
    matched = url.split(/\s+/).includes(val);
  } else {
    matched = url.includes(val);
  }

  if (matched) {
    return {
      text: "",
      attributes: { [attr]: url },
      properties: { [attr]: url },
    };
  }
  return null;
}

export class FingerprintMatcher {
  store: FingerprintStore;

  constructor(store: FingerprintStore) {
    this.store = store;
  }

  match(evidenceData: Partial<CollectedEvidence>): TechMatch[] {
    const headers: Record<string, string> = {};
    for (const [k, v] of Object.entries(evidenceData.headers || {})) {
      headers[k.toLowerCase()] = String(v);
    }

    const cookies: Record<string, string> = {};
    for (const [k, v] of Object.entries(evidenceData.cookies || {})) {
      cookies[k.toLowerCase()] = String(v);
    }

    const meta_tags: Record<string, string> = {};
    for (const [k, v] of Object.entries(evidenceData.meta_tags || {})) {
      meta_tags[k.toLowerCase()] = String(v);
    }

    const script_srcs = Array.isArray(evidenceData.script_srcs) ? evidenceData.script_srcs.map(String) : [];
    const link_hrefs = Array.isArray(evidenceData.link_hrefs) ? evidenceData.link_hrefs.map(String) : [];
    const network_requests = Array.isArray(evidenceData.network_requests)
      ? evidenceData.network_requests.map(req => typeof req === "string" ? req : req.url)
      : [];

    const html_val = (evidenceData as any).html || (evidenceData as any).raw_html || "";
    const rawHtml = String(html_val);

    const js_globals: Record<string, string> = {};
    for (const [k, v] of Object.entries(evidenceData.js_globals || {})) {
      js_globals[_normalize_js_key(k)] = String(v);
    }

    const manifest_url = evidenceData.manifest_url || null;

    const allScripts = [...script_srcs, ...link_hrefs, ...network_requests];

    const domData: Record<string, any[]> = {};
    for (const [k, v] of Object.entries(evidenceData.dom || {})) {
      if (Array.isArray(v)) {
        domData[k] = [...v];
      }
    }

    if (manifest_url) {
      const pwaSel = "link[rel='manifest']";
      if (!domData[pwaSel]) {
        domData[pwaSel] = [
          {
            text: "",
            attributes: { rel: "manifest", href: manifest_url },
            properties: { rel: "manifest", href: manifest_url },
          }
        ];
      }
    }

    const allDomSelectors = this.store.get_all_dom_selectors();
    for (const sel of allDomSelectors) {
      if (!domData[sel]) {
        const findings: any[] = [];
        const subSelectors = sel.split(",").map(s => s.trim());
        for (const subSel of subSelectors) {
          for (const url of network_requests) {
            const finding = _match_network_request_to_selector(url, subSel);
            if (finding) {
              findings.push(finding);
            }
          }
        }
        if (findings.length > 0) {
          domData[sel] = findings;
        }
      }
    }

    const matches: TechMatch[] = [];

    for (const fp of this.store.get_all()) {
      const matched_sources = new Set<string>();
      const evidences: Evidence[] = [];
      const versions: string[] = [];

      // 1. Match Headers
      for (const [headerKey, pattern] of Object.entries(fp.headers)) {
        const headerVal = headers[headerKey.toLowerCase()];
        if (headerVal !== undefined && headerVal !== null) {
          try {
            const rx = new RegExp(pattern, "i");
            const match = rx.exec(headerVal);
            if (match) {
              matched_sources.add("header");
              evidences.push({
                source: "header",
                key: headerKey,
                matched: headerVal,
                pattern: pattern,
              });
              if (match[1]) {
                versions.push(match[1]);
              }
            }
          } catch (e) {
            // Ignored, warning logged in server version
          }
        }
      }

      // 2. Match Cookies
      for (const [cookieKey, pattern] of Object.entries(fp.cookies)) {
        const cookieVal = cookies[cookieKey.toLowerCase()];
        if (cookieVal !== undefined && cookieVal !== null) {
          try {
            const rx = new RegExp(pattern, "i");
            const match = rx.exec(cookieVal);
            if (match) {
              matched_sources.add("cookie");
              evidences.push({
                source: "cookie",
                key: cookieKey,
                matched: cookieVal,
                pattern: pattern,
              });
              if (match[1]) {
                versions.push(match[1]);
              }
            }
          } catch (e) {
            // Ignored
          }
        }
      }

      // 3. Match Meta Tags
      for (const [metaKey, pattern] of Object.entries(fp.meta)) {
        const metaVal = meta_tags[metaKey.toLowerCase()];
        if (metaVal !== undefined && metaVal !== null) {
          try {
            const rx = new RegExp(pattern, "i");
            const match = rx.exec(metaVal);
            if (match) {
              matched_sources.add("meta");
              evidences.push({
                source: "meta",
                key: metaKey,
                matched: metaVal,
                pattern: pattern,
              });
              if (match[1]) {
                versions.push(match[1]);
              }
            }
          } catch (e) {
            // Ignored
          }
        }
      }

      // 4. Match Script Src / Link Hrefs
      for (const pattern of fp.scripts) {
        let matchedScript: string | null = null;
        let matchedVal = "";

        for (const script of allScripts) {
          if (script.includes(pattern)) {
            matchedScript = script;
            matchedVal = script;
            break;
          }
        }

        if (!matchedScript) {
          try {
            const rx = new RegExp(pattern, "i");
            for (const script of allScripts) {
              const match = rx.exec(script);
              if (match) {
                matchedScript = script;
                matchedVal = script;
                if (match[1]) {
                  versions.push(match[1]);
                }
                break;
              }
            }
          } catch (e) {
            // Ignored
          }
        }

        if (matchedScript) {
          matched_sources.add("script");
          evidences.push({
            source: "script",
            key: "script_src",
            matched: matchedVal,
            pattern: pattern,
          });
        }
      }

      // 5. Match HTML
      for (const pattern of fp.html) {
        try {
          const rx = new RegExp(pattern, "i");
          const match = rx.exec(rawHtml);
          if (match) {
            matched_sources.add("html");
            evidences.push({
              source: "html",
              key: "raw_html",
              matched: match[0].slice(0, 100) + "...",
              pattern: pattern,
            });
            if (match[1]) {
              versions.push(match[1]);
            }
          }
        } catch (e) {
          // Ignored
        }
      }

      // 6. Match JS Globals
      for (const [globalKey, pattern] of Object.entries(fp.js_globals)) {
        const normalizedRuleKey = _normalize_js_key(globalKey);
        if (normalizedRuleKey in js_globals) {
          const globalVal = js_globals[normalizedRuleKey];
          let matchedJs = false;

          if (pattern === ".") {
            matchedJs = true;
          } else {
            try {
              const rx = new RegExp(pattern, "i");
              const match = rx.exec(globalVal);
              if (match) {
                matchedJs = true;
                if (match[1]) {
                  versions.push(match[1]);
                }
              }
            } catch (e) {
              // Ignored
            }
          }

          if (matchedJs) {
            matched_sources.add("js_global");
            const truncatedVal = globalVal.length > 100 ? globalVal.slice(0, 100) + "..." : globalVal;
            evidences.push({
              source: "js_global",
              key: globalKey,
              matched: truncatedVal,
              pattern: pattern,
            });
          }
        }
      }

      // 7. Match DOM selectors
      if (fp.dom) {
        if (Array.isArray(fp.dom)) {
          for (const sel of fp.dom) {
            if (sel in domData) {
              matched_sources.add("dom");
              const matchedEl = domData[sel][0];
              let matchedText = `Found element: ${sel}`;
              if (matchedEl && matchedEl.text) {
                matchedText += ` (text: ${matchedEl.text})`;
              }
              evidences.push({
                source: "dom",
                key: "selector",
                matched: matchedText.slice(0, 100),
                pattern: sel,
              });
            }
          }
        } else if (typeof fp.dom === "object") {
          for (const [sel, subRule] of Object.entries(fp.dom)) {
            if (sel in domData) {
              const elements = domData[sel];
              const ruleDict = (subRule && typeof subRule === "object") ? subRule : {};
              if (!subRule || "exists" in ruleDict) {
                matched_sources.add("dom");
                evidences.push({
                  source: "dom",
                  key: sel,
                  matched: `Element exists: ${sel}`,
                  pattern: "",
                });
                continue;
              }

              let matchedAnyEl = false;
              for (const el of elements) {
                let matchedElCriteria = true;

                // Check text regex
                if ("text" in ruleDict) {
                  const pat = ruleDict.text;
                  if (typeof pat === "string") {
                    try {
                      const rx = new RegExp(pat, "i");
                      const match = rx.exec(String(el.text || ""));
                      if (!match) {
                        matchedElCriteria = false;
                      } else {
                        if (match[1]) {
                          versions.push(match[1]);
                        }
                      }
                    } catch (e) {
                      matchedElCriteria = false;
                    }
                  }
                }

                // Check attributes
                if (matchedElCriteria && "attributes" in ruleDict) {
                  const elAttrs = (el.attributes || {}) as Record<string, string>;
                  const attrsRule = ruleDict.attributes;
                  if (attrsRule && typeof attrsRule === "object") {
                    for (const [attrName, pat] of Object.entries(attrsRule)) {
                      const attrVal = elAttrs[attrName];
                      if (attrVal === undefined || attrVal === null) {
                        matchedElCriteria = false;
                        break;
                      }
                      if (typeof pat === "string") {
                        try {
                          const rx = new RegExp(pat, "i");
                          const match = rx.exec(String(attrVal));
                          if (!match) {
                            matchedElCriteria = false;
                            break;
                          } else {
                            if (match[1]) {
                              versions.push(match[1]);
                            }
                          }
                        } catch (e) {
                          matchedElCriteria = false;
                          break;
                        }
                      }
                    }
                  }
                }

                // Check properties
                if (matchedElCriteria && "properties" in ruleDict) {
                  const elProps = (el.properties || {}) as Record<string, string>;
                  const propsRule = ruleDict.properties;
                  if (propsRule && typeof propsRule === "object") {
                    for (const [propName, pat] of Object.entries(propsRule)) {
                      const propVal = elProps[propName];
                      if (propVal === undefined || propVal === null) {
                        matchedElCriteria = false;
                        break;
                      }
                      if (typeof pat === "string") {
                        try {
                          const rx = new RegExp(pat, "i");
                          const match = rx.exec(String(propVal));
                          if (!match) {
                            matchedElCriteria = false;
                            break;
                          } else {
                            if (match[1]) {
                              versions.push(match[1]);
                            }
                          }
                        } catch (e) {
                          matchedElCriteria = false;
                          break;
                        }
                      }
                    }
                  }
                }

                if (matchedElCriteria) {
                  matchedAnyEl = true;
                  const matchedDetails: string[] = [];
                  if ("text" in ruleDict) {
                    matchedDetails.push(`text: ${el.text}`);
                  }
                  if ("attributes" in ruleDict) {
                    const attrsRule = ruleDict.attributes;
                    const attrsDict: Record<string, any> = {};
                    if (attrsRule && typeof attrsRule === "object") {
                      for (const k of Object.keys(attrsRule)) {
                        attrsDict[k] = (el.attributes as any)?.[k];
                      }
                    }
                    matchedDetails.push(`attributes: ${JSON.stringify(attrsDict)}`);
                  }
                  const matchedStr = matchedDetails.length > 0
                    ? `Found: ${sel} (${matchedDetails.join(", ")})`
                    : `Found: ${sel}`;
                  evidences.push({
                    source: "dom",
                    key: sel,
                    matched: matchedStr.slice(0, 120),
                    pattern: JSON.stringify(subRule),
                  });
                  break;
                }
              }

              if (matchedAnyEl) {
                matched_sources.add("dom");
              }
            }
          }
        }
      }

      // Calculate final confidence and build match
      if (evidences.length > 0) {
        const numSources = matched_sources.size;
        let confidence = fp.confidence;

        // +0.1 per additional corroborating source type
        if (numSources > 1) {
          confidence += 0.1 * (numSources - 1);
        }

        // +0.1 if version capture group matched
        const version = versions[0] || null;
        if (version) {
          confidence += 0.1;
        }

        confidence = Math.min(confidence, 1.0);

        matches.push({
          name: fp.name,
          category: fp.category,
          version: version,
          confidence: Number(confidence.toFixed(2)),
          evidence: evidences,
        });
      }
    }

    // Resolve implies chains
    const resolved = this.store.resolve_implies(matches);

    // Sort matches by confidence descending
    resolved.sort((a, b) => b.confidence - a.confidence);

    return resolved;
  }
}
